flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java

```diff
@@ -250,56 +258,64 @@ public void setCurrentDatabase(String databaseName) throws DatabaseNotExistExcep
 	public Optional<CatalogTableOperation> resolveTable(String... tablePath) {
 		checkArgument(tablePath != null && tablePath.length != 0, "Table path must not be null or empty.");
 
-		List<String> defaultPath = new ArrayList<>();
-		defaultPath.add(currentCatalogName);
-		defaultPath.add(currentDatabaseName);
+		List<String> userPath = asList(tablePath);
 
-		List<String> userPath = Arrays.asList(tablePath);
-		defaultPath.addAll(userPath);
+		List<List<String>> prefixes = asList(
+			asList(currentCatalogName, currentDatabaseName),
+			singletonList(currentCatalogName),
+			emptyList()
+		);
 
-		Optional<CatalogTableOperation> inDefaultPath = lookupPath(defaultPath);
-
-		if (inDefaultPath.isPresent()) {
-			return inDefaultPath;
-		} else {
-			return lookupPath(userPath);
+		for (List<String> prefix : prefixes) {
+			Optional<CatalogTableOperation> potentialTable = lookupPath(prefix, userPath);
+			if (potentialTable.isPresent()) {
+				return potentialTable;
+			}
 		}
+
+		return Optional.empty();
 	}
 
-	private Optional<CatalogTableOperation> lookupPath(List<String> path) {
+	private Optional<CatalogTableOperation> lookupPath(List<String> prefix, List<String> userPath) {
 		try {
-			Optional<TableSchema> potentialTable = lookupCatalogTable(path);
+			List<String> path = new ArrayList<>(prefix);
+			path.addAll(userPath);
+
+			Optional<CatalogTableOperation> potentialTable = lookupCatalogTable(path);
```

Comment:
you can check whether the path is of size 3 before calling `lookupCatalogTable()` because we are fully sure that catalog table's full path must be of size 3.

And in `lookupCatalogTable()`, we can change the check from `if (path.size() == 3)` to `checkArgument(path.size() == 3)` and add a comment to let callers be aware that they should make sure their input args are valid

Reply:
External catalogs support arbitrary nesting.

Reply:
External catalogs support arbitrary nesting.

Reply:
External catalogs support arbitrary nesting.

Reply:
which part we are not sure about? isn't table lookup for external catalogs in `lookupExternalTable` rather than `lookupCatalogTable`?

Reply:
what I mean is, `lookupPath()` can be written as 

private Optional<CatalogTableOperation> lookupPath(...) {
	...
	Optional<CatalogTableOperation> potentialTable 
       // if the path is not of size 3, we don't even need to bother with looking it up in catalogs
	if (path.size() == 3) {
		potentialTable = lookupCatalogTable(path);
	}

	if (!potentialTable.isPresent()) {
		potentialTable = lookupExternalTable(path);
	}
	return potentialTable
}

private Optional<CatalogTableOperation> lookupCatalogTable(List<String> path) {
	checkArgument(path.size() == 3, "xxx")
	try {

		...
	} catch () {}
}

Reply:
Ok, I misunderstood you. 

I don't like this approach as this way both `lookupPath` and `lookupCatalogTable` validate structure of the `Catalog` path. In my approach the whole logic is enclosed in a single lookup**CatalogTable** method.

Reply:
the validation is fine. 

> This is not a performance critical code.

I think this will be critical to any benchmarks we run. I just heard from my colleague who is doing Flink/Blink TPCH now that, even how the `CatalogReader` runs is hearting the benchmark quite a bit because it needs to [exam three schema paths](https://github.com/apache/flink/pull/8404/files#diff-0bb30406df09433cce9b3617f910db9bR47) to find the table in a query. So any perf improvement in query planning phase would help us in the long run. Hope the background helps explain the necessity :)

Reply:
the validation is fine. 

> This is not a performance critical code.

Just in general I think query planning is critical to any benchmarks we run. I just heard from my colleague who is doing Flink/Blink TPCH now that, even how the `CatalogReader` runs is hearting the benchmark quite a bit because it needs to [exam three schema paths](https://github.com/apache/flink/pull/8404/files#diff-0bb30406df09433cce9b3617f910db9bR47) to find the table. Hope the background helps explain the necessity. Though I think this probably won't hurt benchmarks run with catalog table and blink planner because a catalog table will be returned without going to look up external tables, any perf improvement in query planning phase would help us in the long run. 

I will leave this issue up to you.

Reply:
I'll leave up to you